### PR TITLE
sasylf lsp implementation with diagnostics and code actions

### PR DESCRIPTION
**Pull request by MonkeyDoug**
#117 
- Adds json interface to sasylf core
- Diagnostics using error messages and severity from sasylf
- Code actions as implemented in eclipse using new classes that replicate eclipse classes

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -9,11 +9,13 @@ VERSION=`head -1 ChangeLog.txt | sed 's/^SASyLF version \(.*\).$$/\1/'`.v`date +
 
 default: test
 
+build : SHELL:=/bin/bash
 build :
 	(cd src && cd edu && cd cmu && cd cs && cd sasylf && cd parser; javacc parser.jj)
 	mkdir -p bin
-	(cd src && javac -source 1.8 -target 1.8 -classpath ../bin:. -d ../bin edu/cmu/cs/sasylf/Main.java edu/cmu/cs/sasylf/term/UnitTests.java)
-	jar cmf sasylf.mf SASyLF.jar ChangeLog.txt -C bin edu -C library org
+	(cd src && javac -cp ../lib/*:../bin:. -source 1.8 -target 1.8 -d ../bin edu/cmu/cs/sasylf/Main.java edu/cmu/cs/sasylf/term/UnitTests.java)
```
Comment (boyland) [inline Makefile:16]: I don't think "\*" in a -cp argument does what you want here.  Globbing works at the shell level; unless bash is doing something strange, "../lib/\*:../bin:." should be looking for a file ../lib/A/bin/:." where A ends with colon and two dots

File: lsp-sample/.eslintignore
Comment (boyland) [inline lsp-sample/.eslintignore:1]: Why is the VSCode extension called "lsp-sample" ?  Why not "lsp-sasylf" ?

File: lsp-sample/client/package.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
+	"name": "lsp-sample-client",
```
Comment (boyland) [inline lsp-sample/client/package.json:2]: Again, why not lsp-sasylf-client?

File: lsp-sample/client/src/extension.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
```
Comment (boyland) [inline lsp-sample/client/src/extension.ts:3]: Is MIT license compatible with creative commons or whatever SASyLF is currently distributed under?

File: src/edu/cmu/cs/sasylf/Main.java [hunk 4]
```diff
@@ -1,13 +1,5 @@
 package edu.cmu.cs.sasylf;
 
-import java.io.File;
-import java.io.FileInputStream;
-import java.io.FileNotFoundException;
-import java.io.IOException;
-import java.io.InputStreamReader;
-import java.io.PrintStream;
-import java.io.Reader;
-
```
Comment (boyland) [inline src/edu/cmu/cs/sasylf/Main.java:10]: It's standard (or was standard) to put all standard imports before imports of app specific classes.  Why reorganize everything?

File: src/edu/cmu/cs/sasylf/Main.java [hunk 5]
```diff
@@ -18,173 +10,242 @@
...
+           * is needed for the quickfixes.
+           */
+          InputStreamReader r;
+          if (args[i].startsWith("--stdin"))
+            r = new InputStreamReader(System.in);
+          else
+            r = new InputStreamReader(new FileInputStream(file), "UTF-8");
+
+          BufferedReader reader = new BufferedReader(r);
+          StringBuilder sb = new StringBuilder();
```
Comment (boyland) [inline src/edu/cmu/cs/sasylf/Main.java:174]: tabs for indentation has the advantage that the viewer can choose what the spacing will be.  You've changed everything to hard spaces.  That not only obscures whether you did anything (did you?) but also means one cannot choose a different indentation view.  Please revert these changes.

**Review by boyland — COMMENTED** [review]
Overall: Lots of good work.  Some specifics:
-- the VS code extension won't be a "sample" extension.  (I had to do the same thing when I updated the Eclipse plugin -- it was all set with Sample this and Test that.)  So update names. Check the licensing too.  (I'm afraid I was lazy about this when I added Eclipse things.)
-- Keep tabs for indentation and revert changes that converted TAB to two spaces
    I wasn't able to tell what you changed in many of the files.
-- An interesting idea to collect common QuickFix things into the core SASyLF, but:
    a. Try not to add a new region system -- we already have one
    b. I'm not sure it was successful.  It seems that only some of the Eclipse quick fixes
        were pulled into common so Eclipse will have to do things over?
    c. It doesn't appear that the Eclipse plugin has been changed to use it.  Rather it appears that some code was copied out (duplicated).  Let's try to do a proper refactoring.
-- the test system that SASyLF has could use some updating, but ignoring it is not the best way forward.  It has some features that it appears that the new system doesn't have or that the new system unaccountably duplicates.

File: src/edu/cmu/cs/sasylf/Main.java (hunk 5 above)
Comment (MonkeyDoug) [reply src/edu/cmu/cs/sasylf/Main.java:190]: Is there a style guide I could reference for the project? I will configure my formatter to follow it. This should solve most of the issues with white spaces, import orders, etc.

File: test.sh
[the remark this answers is not in the document]
Comment (MonkeyDoug) [reply test.sh:1]: I am having some trouble running the tests in the master branch. They work when running "make test" in a fresh clone of the branch. However, after running "make clean" and "make build", "make test" fails with message 
"Error: Could not find or load main class edu.cmu.cs.sasylf.util.UnitTests
Caused by: java.lang.ClassNotFoundException: edu.cmu.cs.sasylf.util.UnitTests"

I also noticed that "edu/cmu/cs/sasylf/term/UnitTests.java" is included in the -d flag in the Makefile. However, the unit tests in the util and reductions directories are not included. Is this on purpose? After I include them in the Makefile the tests run again.

[the remark this answers is not in the document]
Comment (boyland) [reply test.sh:1]: Yes, this was a mistake.  Rohit has a workaround/fix: list more files in the "build" recipe. But these test files should be listed in a definition such as
TESTSRC= \
    edu/cmu/cs/sasylf/term/UnitTests.java \
     edu/cmu/cs/sasylf/util/UnitTests.java \
    edu/cmu/cs/sasylf/reduction/UnitTests.java 

(I'm not sure these are the correct files -- it's from memory.)
And then add ${TESTSRC} after .../Main.java in the compile line.

**Comment by boyland** [comment]
Someone asked about a style guide, but github isn't showing me the comment.

I don't have a written style guide.  It would be good to have one. In it it would say:
1. Use Tabs for indentation.
2. With a control structure: either put the controlled part on the same line, or surround with braces.  In other words, either:
```
if (condition) return 42;
```
or
```
if (condition) {
    return 42;
}
```
Don't write:
```
if (condition)
    return 42;
```
which is subject to confusion.
